lib/HeroBonus.cpp

```diff
@@ -135,18 +144,36 @@ CBonusProxy & CBonusProxy::operator=(CBonusProxy && other)
 	std::swap(bonusListCachedLast, other.bonusListCachedLast);
 	std::swap(selector, other.selector);
 	std::swap(bonusList, other.bonusList);
+	std::swap(currentBonusListIndex, other.currentBonusListIndex);
+
 	return *this;
 }
 
+void CBonusProxy::swapBonusList(TConstBonusListPtr other) const
+{
+	// The idea here is to avoid changing active bonusList while it can be read by a different thread.
+	// Because such use of shared ptr is not thread safe
+	// So to avoid this we change the second offline instance and swap active index
+	auto newCurrent = 1 - currentBonusListIndex;

```

Comment:
Is `currentBonusListIndex` should be only `0` or `1` ?
So, may be it would be more clear to define  `currentBonusListIndex` as bool  => it would be 
`auto newCurrent = !currentBonusListIndex;`
https://stackoverflow.com/questions/2725044/can-i-assume-booltrue-int1-for-any-c-compiler

Reply:
Do not think so